### PR TITLE
[Sema] Emit only "public" access modifier in fix-it for missing witness

### PR DESCRIPTION
**Pull request by rintaro**
```
protocol P {
   func foo()
}
private class C : P {
}
```

Previously, fix-it tries to insert `internal func foo() { <#code#> }`.

With this PR, fix-it emits `public` access modifier only if the protocol and the class are `public` (or `open`).

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by jrose-apple** [comment]
Code LGTM. @benlangmuir?

Can you add a test for a non-type requirement that needs `public` as well?

File: lib/Sema/TypeCheckProtocol.cpp [hunk 1]
```diff
@@ -2032,9 +2032,19 @@ static void diagnoseNoWitness(ValueDecl *Requirement, Type RequirementType,
...
+
+  Accessibility Access = std::min(
+    /* Access of the context */
+    Conformance->getDeclContext()
+      ->getAsGenericTypeOrGenericTypeExtensionContext()->getFormalAccess(),
+    /* Access of the protocol */
+    Requirement->getDeclContext()
+      ->getAsProtocolOrProtocolExtensionContext()->getFormalAccess());
+  if (Access == Accessibility::Public)
+    Printer << "public ";
```
Comment (benlangmuir) [inline lib/Sema/TypeCheckProtocol.cpp]: This needs to get annotated as a keyword.  The easiest way is:
```c++
Printer << tok::kw_public << " ";
```

**Comment by benlangmuir** [comment]
LGTM too

**Comment by rintaro** [comment]
@jrose-apple @benlangmuir Thanks!

@jrose-apple What "non-type requirement" mean? non-class-type?

**Comment by jrose-apple** [comment]
Non-associated-type. Right now the only test that inserts "public" is the associated type one, right? Or are there some that just didn't show up in the diff because they didn't change?

File: test/decl/protocol/conforms/fixit_stub.swift [hunk 2]
```diff
@@ -75,9 +75,54 @@ extension Adopter6: ProtocolWithSelfRequirement2 { //expected-error{{type 'Adopt
...
+public protocol ProtocolWithPublicAccess2 {
+  associatedtype AssocType //expected-note{{protocol requires nested type 'AssocType'}} {71-71=\n    typealias AssocType = <#type#>\n}}
+}
+class Adopter8: ProtocolWithPublicAccess1, ProtocolWithPublicAccess2 {
+  // expected-error@-1{{type 'Adopter8' does not conform to protocol 'ProtocolWithPublicAccess1'}}
+  // expected-error@-2{{type 'Adopter8' does not conform to protocol 'ProtocolWithPublicAccess2'}}
+}
+
+public protocol ProtocolWithPublicAccess3 {
+  func foo() // expected-note{{protocol requires function 'foo()' with type '() -> ()'; do you want to add a stub?}} {{78-78=\n    public func foo() {\n        <#code#>\n    \}\n}}
```
Comment (rintaro) [inline test/decl/protocol/conforms/fixit_stub.swift]: @jrose-apple Here's non-`associatedtype` test. Do we need `init()`, `var`, or `subscript` also?

Comment (rintaro) [inline test/decl/protocol/conforms/fixit_stub.swift]: :)
Sorry, I'm not authorized to merge. Could you?

Or @shahmishal, Could you authorize me to merge?

Comment (benlangmuir) [inline test/decl/protocol/conforms/fixit_stub.swift]: Instructions for getting commit access are at https://swift.org/contributing/#contributing-code "Commit Access" 😃

**Comment by rintaro** [comment]
@benlangmuir Thanks!

Actually I've been granted commit access on Jun. 1.
IIRC, from around [Jul. 27](https://lists.swift.org/pipermail/swift-dev/Week-of-Mon-20160725/002512.html), it seems only Apple members can merge into the `master`.
We non-apple "granted" contributors can trigger CI, but can't push Github "Merge pull request" buttons on this repo.
@modocache or @jtbandes, Could you confirm this?

**Comment by modocache** [comment]
Yes, I believe that's true. I'm unable to merge my code directly.

**Comment by jtbandes** [comment]
Same here, at least in this repo. I can merge into swift-evolution.

**Comment by benlangmuir** [comment]
@shahmishal tells me this should be fixed now.  After the required tests have passed, you should be able to hit the merge button if you have commit access.  Let us know if it doesn't work!

**Comment by rintaro** [comment]
Confirmed on other PRs. Thank you!

**Comment by jtbandes** [comment]
Yes, the merge button is green now. Thanks 😄

**Pull request merged**